src/nor_flash.rs

```diff
@@ -0,0 +1,239 @@
+use crate::{iter::IterableByOverlaps, ReadStorage, Region, Storage};
+
+/// Read only NOR flash trait.
+pub trait ReadNorFlash {
+	/// An enumeration of storage errors
+	type Error;
+
+	/// The minumum number of bytes the storage peripheral can read
+	const READ_SIZE: usize;
+
+	/// Read a slice of data from the storage peripheral, starting the read
+	/// operation at the given address offset, and reading `bytes.len()` bytes.
+	///
+	/// This should throw an error in case `bytes.len()` will be larger than
+	/// the peripheral end address.
+	fn try_read(&mut self, offset: u32, bytes: &mut [u8]) -> Result<(), Self::Error>;
+
+	/// The capacity of the peripheral in bytes.
+	fn capacity(&self) -> usize;
+}
+
+/// NOR flash trait.
+pub trait NorFlash: ReadNorFlash {
+	/// The minumum number of bytes the storage peripheral can write
+	const WRITE_SIZE: usize;
+
+	/// The minumum number of bytes the storage peripheral can erase
+	const ERASE_SIZE: usize;
+
+	/// Erase the given storage range, clearing all data within `[from..to]`.
+	/// The given range will contain all 1s afterwards.
+	///
+	/// This should return an error if the range is not aligned to a proper
+	/// erase resolution
+	/// Erases page at addr, sets it all to 0xFF
+	/// If power is lost during erase, contents of the page are undefined.
+	/// `from` and `to` must both be multiples of `erase_size()` and `from` <= `to`.
+	fn try_erase(&mut self, from: u32, to: u32) -> Result<(), Self::Error>;
+
+	/// Writes data to addr, bitwise ANDing if there's already data written at that location,
+	/// If power is lost during write, the contents of the written words are undefined.
+	/// The rest of the page is guaranteed to be unchanged.
+	/// It is not allowed to write to the same word twice.
+	/// `offset` and `bytes.len()` must both be multiples of `write_size()` and properly aligned.
```

Comment:
WRITE_SIZE

"and properly aligned" is redundant, it's implied by the "multiple of" restriction